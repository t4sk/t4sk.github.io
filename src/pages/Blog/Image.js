import React from "react"
import PropTypes from "prop-types"
import { Image as SemanticImage } from "semantic-ui-react"
import "./Image.css"

function Image(props) {
  const { src, alt } = props

  return (
    <div className="Image">
      <SemanticImage src={src} alt={alt} />
    </div>
  )
}

Image.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
}

export default Image
