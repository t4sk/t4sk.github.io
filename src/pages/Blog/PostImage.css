.PostImage {
  display: flex;
  flex-direction: row;
  justify-content: center;
}
