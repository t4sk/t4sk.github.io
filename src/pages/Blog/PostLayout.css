.PostLayout {
  padding: 10px;
  margin-bottom: 30px;
}
.PostLayout p {
  font-size: 16px;
}
