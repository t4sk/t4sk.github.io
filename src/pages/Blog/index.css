.Blog-list {
  list-style: none;
}
.Blog-list li {
  line-height: 1.75em;
}
.Blog-list .Blog-date {
  color: rgba(0, 0, 0, 0.5);
  margin-right: 7px;
}
