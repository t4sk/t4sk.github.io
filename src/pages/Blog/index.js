import React from "react"
import moment from "moment"
import PropTypes from "prop-types"
import { Tab, Segment } from "semantic-ui-react"
import Layout from "../../components/Layout"
import SEO from "../../components/SEO"
import "./index.css"
import LANG from "./index.lang"
import translate from "../../translate"

// import Posts from "./posts-bk"
import Posts from "./posts"

export function groupByYear(posts) {
  const grouped = posts.reduce((grouped, post) => {
    const year = moment(post.date).year()

    if (grouped[year]) {
      grouped[year].push(post)
    } else {
      grouped[year] = [post]
    }

    return grouped
  }, {})

  const years = Object.keys(grouped).sort((a, b) => parseInt(b) - parseInt(a))

  // sort posts by date desc
  for (let year of years) {
    grouped[year].sort((a, b) => moment(b.date).unix() - moment(a.date).unix())
  }

  return {
    years,
    byYear: grouped,
  }
}

function PostsByYear(props) {
  const { posts } = props

  const { years, byYear } = groupByYear(posts)

  return years.map((year, i) => (
    <div className="Blog-posts-by-year" key={i}>
      <h3>{year}</h3>
      <ul className="Blog-list">
        {byYear[year].map((post, j) => (
          <li key={j}>
            <div className="Blog-post">
              <div className="Blog-date">{post.date}</div>
              <a href={`/${post.lang}/blog/${post.date}`}>{post.title}</a>
            </div>
          </li>
        ))}
      </ul>
    </div>
  ))
}

const panes = [
  {
    menuItem: "English",
    render: () => (
      <PostsByYear posts={Posts.filter(post => post.lang == "en")} />
    ),
  },
  {
    menuItem: "Japanese",
    render: () => (
      <PostsByYear posts={Posts.filter(post => post.lang == "jp")} />
    ),
  },
]

const ACTIVE_INDEX = {
  en: 0,
  jp: 1,
}

function Blog(props) {
  const { lang } = props

  const title = `${translate(LANG, lang, "Blog")} | Smart Contract Programmer`

  return (
    <Layout lang={lang}>
      <SEO title={title} lang={lang} />
      <Segment>
        <Tab
          menu={{ secondary: true }}
          defaultActiveIndex={ACTIVE_INDEX[lang]}
          panes={panes}
        />
      </Segment>
    </Layout>
  )
}

Blog.propTypes = {
  lang: PropTypes.oneOf(["en", "jp"]).isRequired,
}

export default Blog
