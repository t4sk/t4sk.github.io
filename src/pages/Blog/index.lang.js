const LANG = {
  en: {
    Blog: "Blog",
  },
  jp: {
    Blog: "ブログ",
  },
}

export default LANG
