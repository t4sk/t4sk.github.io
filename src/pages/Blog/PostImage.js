import React from "react"
import PropTypes from "prop-types"
import { Image } from "semantic-ui-react"
import "./PostImage.css"

function PostImage(props) {
  const { src, alt } = props

  return (
    <span className="PostImage">
      <Image src={src} alt={alt} bordered />
    </span>
  )
}

PostImage.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
}

export default PostImage
