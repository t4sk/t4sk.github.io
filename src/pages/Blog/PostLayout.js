import React from "react"
import PropTypes from "prop-types"
import { Segment } from "semantic-ui-react"
import Layout from "../../components/Layout"
import SEO from "../../components/SEO"
import "./PostLayout.css"

function PostLayout(props) {
  const { lang, title, date, children } = props

  return (
    <Layout lang={lang}>
      <SEO title={`${title} | Smart Contract Programmer`} lang={lang} />
      <Segment>
        <div className="PostLayout">
          <h1>{title}</h1>
          <p>
            <em>{date}</em>
          </p>
          <hr />

          {children}
        </div>
      </Segment>
    </Layout>
  )
}

PostLayout.propTypes = {
  lang: PropTypes.oneOf(["en", "jp"]).isRequired,
  title: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
}

export default PostLayout
