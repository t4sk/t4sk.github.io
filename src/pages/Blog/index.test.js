import React from "react"
import { shallow } from "enzyme"
import Blog, { groupByYear } from "./index"

const props = {
  lang: "en",
}

test("en", () => {
  const component = shallow(<Blog lang="en" />)

  expect(component).toMatchSnapshot()
})

test("jp", () => {
  const component = shallow(<Blog lang="jp" />)

  expect(component).toMatchSnapshot()
})

test("groupByYear", () => {
  expect(groupByYear(POSTS)).toEqual({
    years: ["2019", "2018", "2017"],
    byYear: {
      "2019": [POSTS[0], POSTS[1]],
      "2018": [POSTS[2], POSTS[3], POSTS[4]],
      "2017": [POSTS[5]],
    },
  })
})

const POSTS = [
  {
    title: "Introduction to Ethereum Payment Channels Video",
    date: "2019-01-03",
    lang: "en",
  },
  {
    title: "Implementing Harberger Tax Deeds",
    date: "2019-01-02",
    lang: "en",
  },
  {
    title: "Writing a Dollar Auction Contract",
    date: "2018-10-11",
    lang: "en",
  },
  {
    title: "Writing a Penny Auction Contract",
    date: "2018-10-10",
    lang: "en",
  },
  {
    title: "Contracts Calling Arbitrary Functions",
    date: "2018-10-01",
    lang: "en",
  },
  {
    title: "Testing and Deploying Smart Contracts with Remix",
    date: "2017-12-12",
    lang: "en",
  },
]
