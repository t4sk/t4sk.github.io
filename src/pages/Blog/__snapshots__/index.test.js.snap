// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`en 1`] = `
<Layout
  lang="en"
>
  <SEO
    description="Learn how to program smart contracts."
    keywords={
      Array [
        "smart contract",
        "solidity",
        "ethereum",
      ]
    }
    lang="en"
    meta={Array []}
    title="Blog | Smart Contract Programmer"
  />
  <PostsByYear
    posts={
      Array [
        Object {
          "component": [Function],
          "date": "2019-04-11",
          "lang": "en",
          "title": "Solidity by Example - Part 0 - Setting Up MetaMask",
        },
      ]
    }
  />
</Layout>
`;

exports[`jp 1`] = `
<Layout
  lang="jp"
>
  <SEO
    description="Learn how to program smart contracts."
    keywords={
      Array [
        "smart contract",
        "solidity",
        "ethereum",
      ]
    }
    lang="jp"
    meta={Array []}
    title="ブログ | Smart Contract Programmer"
  />
  <PostsByYear
    posts={
      Array [
        Object {
          "component": [Function],
          "date": "2019-04-11",
          "lang": "jp",
          "title": "Solidity 入門 - パート 0 - MetaMaskの設定",
        },
      ]
    }
  />
</Layout>
`;
